### PR TITLE
ngx ssl add client addr function

### PR DESCRIPTION
**Pull request by smartwjw**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_ssl_certby.c [hunk 1]
```diff
@@ -880,6 +880,30 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+    }
+
+    ssl_conn = r->connection->ssl->connection;
+    if (ssl_conn == NULL) {
+        *err = "bad ssl conn";
+        return NGX_ERROR;
+    }
+
+    addr_text = &r->connection->addr_text;
+    *addr = (char *)addr_text->data;
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Style: need a space after `(char *)`.

File: src/ngx_http_lua_ssl_certby.c [hunk 2]
```diff
@@ -880,6 +880,30 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
 #endif
 }
 
+int
+ngx_http_lua_ffi_ssl_client_addr(ngx_http_request_t *r, char **addr,
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: OK, you are returning a raw client address here, so you should rename this C function to something like `ngx_http_lua_ffi_ssl_raw_client_addr`, just like the existing `ngx_http_lua_ffi_ssl_raw_server_addr` C function. We should be consistent with ourselves here.

Comment (smartwjw) [inline src/ngx_http_lua_ssl_certby.c]: I first name it `raw_client_addr`, I am returning ip address in plain text, like `ngx.var.remote_addr`, not `ngx.var.binary_remote_addr`, so I name it `client_addr`, am I correct?

File: src/ngx_http_lua_ssl_certby.c (hunk 1 above)
Comment (smartwjw) [inline src/ngx_http_lua_ssl_certby.c]: I'll fix it soon.

File: src/ngx_http_lua_ssl_certby.c (hunk 2 above)
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: @smartwjw I see. Need some test cases for various different cases like unix domain socket clients.

Comment (smartwjw) [inline src/ngx_http_lua_ssl_certby.c]: @agentzh I've added some test cases for `client_addr`.

File: t/139-ssl-cert-by.t [hunk 3]
```diff
@@ -1866,3 +1866,198 @@ failed to do SSL handshake: handshake failed
...
+close: 1 nil
+
+--- error_log
+client ip: 127.0.0.1
+
+--- no_error_log
+[error]
+[alert]
+
+
```
Comment (agentzh) [inline t/139-ssl-cert-by.t]: Please use the [reindex](https://github.com/openresty/openresty-devel-utils/blob/master/reindex) script to re-format this `.t` file. Thank you!

File: t/139-ssl-cert-by.t [hunk 4]
```diff
@@ -1866,3 +1866,198 @@ failed to do SSL handshake: handshake failed
...
+received: Server: nginx
+received: Content-Type: text/plain
+received: Content-Length: 4
+received: Connection: close
+received: 
+received: foo
+close: 1 nil
+
+--- error_log
+client ip:
```
Comment (agentzh) [inline t/139-ssl-cert-by.t]: We should support unix socket clients, no?

Comment (smartwjw) [reply t/139-ssl-cert-by.t]: Yes,  but unix socket address is useless in this scenery, so I simply return empty string when using unix socket, is that OK?

File: t/139-ssl-cert-by.t (hunk 3 above)
Comment (smartwjw) [reply t/139-ssl-cert-by.t]: done

File: t/139-ssl-cert-by.t (hunk 4 above)
Comment (agentzh) [reply t/139-ssl-cert-by.t]: @smartwjw No, it's not OK. This API should be compatible with UNIX domain socket. And a single nginx `server {}` may serve different kinds of incoming SSL traffic at the same time. Being able to inspect unix domain sockets is also very useful. And we should be consistent with the existing `raw_server_addr` API.

Comment (agentzh) [reply t/139-ssl-cert-by.t]: @smartwjw For such a low level API, we should not only target your own single use case. That's just not right :) Thanks for your understanding!

File: src/ngx_http_lua_ssl_certby.c (hunk 2 above)
Comment (agentzh) [reply src/ngx_http_lua_ssl_certby.c]: I think we should do raw_client_addr here, which is more useful. Parsing the binary form is much more efficient than parsing the textual form.

File: src/ngx_http_lua_ssl_certby.c [hunk 5]
```diff
@@ -880,6 +880,36 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
 #endif
 }
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Style: nginx requires 2 blank lines to separate adjacent C function definitions.

File: src/ngx_http_lua_ssl_certby.c [hunk 6]
```diff
@@ -880,6 +880,36 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+    *addrlen = 0;
+    return NGX_OK;
+#endif
+
+    addr_text = &r->connection->addr_text;
+    *addr = (char *) addr_text->data;
+    *addrlen = addr_text->len;
+
+    return NGX_OK;
+}
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Ditto.

File: t/139-ssl-cert-by.t [hunk 7]
```diff
@@ -1866,3 +1866,199 @@ failed to do SSL handshake: handshake failed
...
+received: Server: nginx
+received: Content-Type: text/plain
+received: Content-Length: 4
+received: Connection: close
+received: 
+received: foo
+close: 1 nil
+
+--- error_log
+client addr:
```
Comment (agentzh) [inline t/139-ssl-cert-by.t]: Can we get the address of the client unix domain socket at all?

File: src/ngx_http_lua_ssl_certby.c [hunk 8]
```diff
@@ -880,6 +880,72 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+    size_t *addrlen, int *addrtype, char **err)
+{
+#if (NGX_HAVE_UNIX_DOMAIN)
+    struct sockaddr_un *saun;
+#endif
+    ngx_ssl_conn_t *ssl_conn;
+    ngx_connection_t *c;
+    struct sockaddr_in *sin;
+#if (NGX_HAVE_INET6)
+    struct sockaddr_in6 *sin6;
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Style: all the local variable identifiers should be aligned up vertically. Please see other C functions for examples.

File: src/ngx_http_lua_ssl_certby.c (hunk 5 above)
Comment (agentzh) [reply src/ngx_http_lua_ssl_certby.c]: @smartwjw Seen this comment?

File: src/ngx_http_lua_ssl_certby.c [hunk 9]
```diff
@@ -880,6 +880,72 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+
+        break;
+#endif
+
+# if(NGX_HAVE_UNIX_DOMAIN)
+    case AF_UNIX:
+        saun = (struct sockaddr_un *)c->sockaddr;
+        /* on Linux sockaddr might not include sun_path at all */ 
+        if(c->socklen <= (socklen_t) offsetof(struct sockaddr_un, sun_path))
+        {
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Style: `{` should be in the previous line unless the `if` condition is too long and would exceed 80 columns.

File: src/ngx_http_lua_ssl_certby.c [hunk 10]
```diff
@@ -880,6 +880,72 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+        *addrtype = NGX_HTTP_LUA_ADDR_TYPE_INET6;
+
+        break;
+#endif
+
+# if(NGX_HAVE_UNIX_DOMAIN)
+    case AF_UNIX:
+        saun = (struct sockaddr_un *)c->sockaddr;
+        /* on Linux sockaddr might not include sun_path at all */ 
+        if(c->socklen <= (socklen_t) offsetof(struct sockaddr_un, sun_path))
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Style: need a space after the keyword `if`. Please see the rest of the code base (and also the official nginx core) for examples.

File: src/ngx_http_lua_ssl_certby.c [hunk 11]
```diff
@@ -880,6 +880,72 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+        {
+            *addr = "";
+            *addrlen = 0;
+        } else {
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Style: need a blank line before this `} else {` line. Please see the rest of the code base (and also the official nginx core) for examples.

File: t/139-ssl-cert-by.t (hunk 7 above)
Comment (agentzh) [reply t/139-ssl-cert-by.t]: Do we support unix domain socket now? Hello?

**Comment by agentzh** [comment]
@smartwjw BTW, using the `ngx-releng` script found the following coding style issues in your code:

screen shot 2017-10-06 at 1 04 21 pm

File: src/ngx_http_lua_ssl_certby.c [hunk 12]
```diff
@@ -880,6 +880,72 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+        *addr = (char *) &sin6->sin6_addr.s6_addr;
+        *addrtype = NGX_HTTP_LUA_ADDR_TYPE_INET6;
+
+        break;
+#endif
+
+# if(NGX_HAVE_UNIX_DOMAIN)
+    case AF_UNIX:
+        saun = (struct sockaddr_un *)c->sockaddr;
+        /* on Linux sockaddr might not include sun_path at all */
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Line trailing space on this line. Also detected by the `ngx-releng` tool.

**Comment by agentzh** [comment]
Also, I noticed that your branch fails the travis ci test. Will you make that pass? BTW, you can point lua-resty-core to your own branch in `.travis.yml` temporarily in this branch.

File: t/139-ssl-cert-by.t (hunk 7 above)
Comment (smartwjw) [reply t/139-ssl-cert-by.t]: The program itself support unix domain now, but I tested on my mac and a CentOS vm, it all returns empty string, so the test code don't need to change.

File: src/ngx_http_lua_ssl_certby.c [hunk 13]
```diff
@@ -881,6 +881,74 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
...
+        *addrlen = 16;
+        *addr = (char *) &sin6->sin6_addr.s6_addr;
+        *addrtype = NGX_HTTP_LUA_ADDR_TYPE_INET6;
+
+        break;
+#endif
+
+# if (NGX_HAVE_UNIX_DOMAIN)
+    case AF_UNIX:
+        saun = (struct sockaddr_un *)c->sockaddr;
```
Comment (agentzh) [inline src/ngx_http_lua_ssl_certby.c]: Style: needs a space after type casting expressions like `(struct sockaddr_un *)`.

File: t/139-ssl-cert-by.t [hunk 14]
```diff
@@ -1866,3 +1866,206 @@ failed to do SSL handshake: handshake failed
...
+
+    server {
+        listen 127.0.0.1:12345 ssl;
+        server_name   test.com;
+
+        ssl_certificate_by_lua_block { 
+            local ssl = require "ngx.ssl"
+            local byte = string.byte
+            local addr, addrtype, err = ssl.raw_client_addr()
+            local ip = string.format("%d.%d.%d.%d", byte(addr, 1), byte(addr, 2),
```
Comment (spacewander) [inline t/139-ssl-cert-by.t]: @smartwjw 
Better to localize `string.format` just like the `string.byte`?

File: t/139-ssl-cert-by.t [hunk 15]
```diff
@@ -1866,3 +1866,206 @@ failed to do SSL handshake: handshake failed
...
+    lua_package_path "../lua-resty-core/lib/?.lua;;";
+
+    server {
+        listen 127.0.0.1:12345 ssl;
+        server_name   test.com;
+
+        ssl_certificate_by_lua_block { 
+            local ssl = require "ngx.ssl"
+            local byte = string.byte
+            local addr, addrtype, err = ssl.raw_client_addr()
```
Comment (spacewander) [inline t/139-ssl-cert-by.t]: Better to check if `addrtype` is equal to `inet`.

File: t/139-ssl-cert-by.t [hunk 16]
```diff
@@ -1866,3 +1866,206 @@ failed to do SSL handshake: handshake failed
...
+--- http_config
+    lua_package_path "../lua-resty-core/lib/?.lua;;";
+
+    server {
+        listen unix:$TEST_NGINX_HTML_DIR/nginx.sock ssl;
+        server_name   test.com;
+
+        ssl_certificate_by_lua_block { 
+            local ssl = require "ngx.ssl"
+            local addr, addrtyp, err = ssl.raw_client_addr()
```
Comment (spacewander) [inline t/139-ssl-cert-by.t]: Better to check if `addrtyp` is equal to `unix`.

File: t/139-ssl-cert-by.t [hunk 17]
```diff
@@ -1866,3 +1866,206 @@ failed to do SSL handshake: handshake failed
...
+received: 
+received: foo
+close: 1 nil
+
+--- error_log
+client socket file: 
+
+--- no_error_log
+[error]
+[alert]
```
Comment (spacewander) [inline t/139-ssl-cert-by.t]: Is it possible to add a test case for IPv6, like the [TEST 54: IPv6](https://github.com/openresty/lua-nginx-module/blob/master/t/058-tcp-socket.t#L3231)?

Simply check if the client ip and addrtype are expected is enough.

File: t/139-ssl-cert-by.t [hunk 18]
```diff
@@ -1866,3 +1866,206 @@ failed to do SSL handshake: handshake failed
...
+
+=== TEST 22: get raw_client_addr - IPv4
+--- http_config
+    lua_package_path "../lua-resty-core/lib/?.lua;;";
+
+    server {
+        listen 127.0.0.1:12345 ssl;
+        server_name   test.com;
+
+        ssl_certificate_by_lua_block {
```
Comment (agentzh) [inline t/139-ssl-cert-by.t]: Should remove the line-trailing spaces on this line.

File: t/139-ssl-cert-by.t [hunk 19]
```diff
@@ -1866,3 +1866,206 @@ failed to do SSL handshake: handshake failed
...
+
+=== TEST 23: get raw_client_addr - unix domain socket
+--- http_config
+    lua_package_path "../lua-resty-core/lib/?.lua;;";
+
+    server {
+        listen unix:$TEST_NGINX_HTML_DIR/nginx.sock ssl;
+        server_name   test.com;
+
+        ssl_certificate_by_lua_block {
```
Comment (agentzh) [inline t/139-ssl-cert-by.t]: Ditto.

**Comment by agentzh** [comment]
Merged with minor fixes. Thanks!

**Pull request closed without merging**